src/lib.rs

```diff
@@ -423,67 +451,147 @@ impl<A: Arbitrary> Arbitrary for ::std::num::Wrapping<A> {
     }
 }
 
+/// An enumeration of buffer creation errors
+#[derive(Debug, Clone, Copy)]
+pub enum BufferError {
+    /// The input buffer is empty, causing construction of some buffer types to
+    /// fail
+    EmptyInput,
+}
+
+/// A source of unstructured data with a finite size
+///
+/// This buffer is a finite source of unstructured data. Once the data is
+/// exhausted it stays exhausted.
+pub struct FiniteBuffer<'a> {
+    buffer: &'a [u8],
+    offset: usize,
+    max_len: usize,
+}
+
+impl<'a> FiniteBuffer<'a> {
+    /// Create a new FiniteBuffer
+    ///
+    /// If the passed `buffer` is shorter than max_len the total number of bytes
+    /// will be the bytes available in `buffer`. If `buffer` is longer than
+    /// `max_len` the buffer will be trimmed.
+    pub fn new(buffer: &'a [u8], max_len: usize) -> Result<Self, BufferError> {
+        let buf: &'a [u8] = if buffer.len() > max_len {
+            &buffer[..max_len]
+        } else {
+            // This branch is hit if buffer is shorter than max_len. We might
+            // choose to make this an error condition instead of, potentially,
+            // surprising folks with less bytes.
+            buffer
+        };
+
+        Ok(FiniteBuffer {
+            buffer: buf,
+            offset: 0,
+            max_len: buf.len(),
+        })
+    }
+}
+
+impl<'a> Unstructured for FiniteBuffer<'a> {
+    type Error = ();
+
+    fn fill_buffer(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        if (self.max_len - self.offset) >= buffer.len() {
+            let max = self.offset + buffer.len();
+            for (i, idx) in (self.offset..max).enumerate() {
+                buffer[i] = self.buffer[idx];
+            }
+            self.offset = max;
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    // NOTE(blt) I'm not sure if this is the right definition. I don't
+    // understand the purpose of container_size.
+    fn container_size(&mut self) -> Result<usize, Self::Error> {
```

Comment:
`container_size` is intended to be used as a function to generate a size for some container of data. Consider, for example, an `arbitrary` implementation for `Vec` will first generate the vector length using `container_size` and then fill the data using `fill_buffer`.

The intention here is to avoid generating extremely large containers that are simply a view of the same ever-repeating sequence of bytes.